Public API (#85)
astropy/astropy-APEs: APE22.rst

> Astropy's Public API Policy

Author: ```suggestion
Astropy's API Policy on Modules and their Contents
```

---------------------------

author: Nathaniel Starkman

date-created: 2013 November 5 <replace with the date you submit the APE>

date-last-revised: 2013 December 17 <keep this up to date anytime something
changes>

date-accepted: 2014 January 20 <replace with accepted date>

type: <one of these three: Standard Track, Informational, Process>

status: Discussion

Abstract
--------

This APE establishes Astropy's policy for public versus internal modules and
their contents.

Scope
-----

A library's API is the publicly accessible "surface": the set of modules,
classes, functions, and other symbols provided to the library's users. This APE
is narrowly focused on the module level of the API -- what modules are public
and what modules are internal -- and the objects within those modules -- which

> are public and how this is communicated. This APE does not address any other
> questions of API design, e.g. public versus internal methods in classes.

Author: ```suggestion
are public and how this is communicated. This APE does not address any other
questions of API design, e.g. public versus internal methods in classes, leaving
these for future APEs or future revisions to this APE.
```

- Is ``qux`` public? It does not start with an underscore and is in
  ``baz.__all__``, but what if ``baz`` is private?

Astropy partially implements all three of the conventions: some private classes
and functions are prefixed with an underscore, while some are not; some modules
have an ``__all__`` attribute, and some do not;  some (debatably) internal
interfaces are in the public documentation, like ``astropy.utils``, and some are
not. However, Astropy does not have a uniform approach to differentiating and
communicating public vs internal symbols, which leads to confusion for users and
developers alike. This longstanding issue has steadily become more pressing with
the development of assistive code tools, like auto-fill in IDEs or more recently
with generative coding tools, like GitHub's CoPilot, which suggest code to users
based on a corpus of code examples.  When users unintentionally use internal
code, code tools will then suggest that internal code to other users. If we do
not have a uniform and systematic approach to communicating what is public vs
internal, then we cannot expect users and especially their tools to know what is
public vs internal.

> 

Author: From https://github.com/astropy/astropy/issues/15169, I recall another source of issues that has bitten astropy in the past: I/O registration. When deep sub-modules look public, e.g. `astropy.cosmology.flrw.lambdacdm.LambdaCDM` then that is generally used as the class pathway in I/O rather than the better `astropy.cosmology.LambdaCDM`. When the class is moved it breaks the I/O and needs a whole backwards-compatibility logic to support files using the previous path. Cleaning up public vs internal, e.g. to `astropy.cosmology._flrw.lambdacdm.LambdaCDM` would make this an obviously bad way to serialize the class and code authors would get it "right" (`astropy.cosmology.LambdaCDM`) the first time. Clear public API makes for more stable code.

probably deserves a few sentences.

Reviewer: This is slightly orthogonal to the issue here: really, one needs to set `__module__` to avoid this - registration just followed that, not a choice made by coders. Note that `numpy` in fact does this - which partially is annoying, since one no longer knows where the function is defined.

Author: For me the pain point was YAML serialization because I used ``f"{class.__module__}.{class.__qualname__}"``, which looks reasonable as code, but led to the problems described above. I would not have pushed the code as I originally did if, when making the tests, I had noticed a private path in the way.

Reviewer: Yes, that one is an issue for examples and `sphinx` too - I really dislike that when we made `representations` a module (which was definitely a good idea!), we had to make so many updates to the docs. It would also be nice to use `regex` for things like [`__construct_mixin_classes`](https://github.com/astropy/astropy/blob/291414a3b600a12738762b7c50bfc6709f3cbf13/astropy/table/serialize.py#L25)

For me, this is still separate from the APE, since this has annoyed me amply already! But I can see that coming from an environment where files with leading underscores have more meaning, you could have avoided the issues.

Author: Agreed we can leave it out of this APE. I think the point I made "If, when making the tests, I had noticed a private path I would have fixed the problem from the start." means adopting this APE will automatically force most of these I/O issues to be fixed

4. **Add a pre-commit bot**. This will ensure that the ``__all__`` attribute is
   always present and up-to-date. It will also ensure that the documentation is
   always up-to-date with the ``__all__`` attribute. This can be accomplished
   with a pre-commit hook that runs a script that checks the ``__all__``
   attribute and, if the module is public, searches in the ``docs/api``
   directory for symbols in ``__all__``.

5. **Add prefixes**: Add prefixes to the top-most private symbols. For modules
   this makes all their contents private.

In Astropy core each phase will be implemented on a per-sub-package basis with
individual pull requests.

Backward compatibility
----------------------

This APE breaks backward compatibility in two ways:

1. Changes ``__all__`` in many modules. Many modules define an ``__all__`` but
   include symbols that are intended to be private because they should be
   imported from other modules (generally the top of the module). This will not
   break code that directly imports the symbols (as Python does not use
   ``__all__`` for this purpose), but it will break code that expects private
   symbols and uses ``import *``.

> 2. Adds prefixes to many objects that are not public. This can be done in a

Author: ```suggestion
2. Adds prefixes to many symbols that are not public. This can be done in a
```

   backward compatible way by adding a ``__getattr__`` method to the module that
   raises a warning for any object that is not public. This will allow existing
   code to continue to work, while encouraging people to fix their code to use
   the public interface.

Alternatives
------------

**We do nothing:**

This is the status quo.  It is not a good option because it is not solve the
issue.  The aforementioned problems of not i) knowing what is stable and
supported, and what is not, remain.

**We allow** ``__all__`` **to be optional:**

This is not great.

The only time this might be good is when a module has dynamic symbols from a
`PEP 562 <https://peps.python.org/pep-0562/>`_ module-level ``__getattr__``
method. However, if it truly dynamic then it cannot be statically analyzed, which
is undesirable for other reasons. If it is not truly dynamic, then the "dynamic"
symbols can be added to ``__all__`` and an ``__init__.pyi`` file used to
communicate the public interface to static analyzers.

**We make the documentation the authoritative definition of the API.**

> Good code is well documented. Beyond code comments and docstrings, what is

Author: ```suggestion
Good code is well documented in the code. Beyond code comments and docstrings, what is
```

Astropy's API Policy on Modules and their Contents
---------------------------

author: Nathaniel Starkman

> date-created: 2013 November 5 <replace with the date you submit the APE>

Reviewer 2: I guess this date can be set already ?

date-last-revised: 2013 December 17 <keep this up to date anytime something
changes>

date-accepted: 2014 January 20 <replace with accepted date>

type: <one of these three: Standard Track, Informational, Process>

status: Discussion

Abstract
--------

This APE establishes Astropy's policy for public versus internal modules and
their contents.

Scope
-----

A library's API is the publicly accessible "surface": the set of modules,
classes, functions, and other symbols provided to the library's users. This APE
is narrowly focused on the module level of the API -- what modules are public
and what modules are internal -- and the objects within those modules -- which
are public and how this is communicated. This APE does not address any other
questions of API design, e.g. public versus internal methods in classes, leaving
these for future APEs or future revisions to this APE.

.. _description_of_the_problem:

Description of the Problem
--------------------------

Pure Python has no language-level means to establish and enforce separations
between public versus internal interfaces -- all code objects can be accessed.

> This is in contrast to languages like C++ and Java, which have language-level

Reviewer 2: very minor tweak: the "and" makes it sound like those are the only two languages with a private VS public boundary :)
```suggestion
This is in contrast to languages like C++ or Java, which have language-level
```

constructs to establish public versus private interfaces. Differentiating
between public and internal interfaces is important for everyone, users and
developers alike, to know what interfaces are stable and supported, and which
are not. Consequently, Python has developed conventions for communicating what
is public versus internal. Nominally there are three means (discussed more
formally later):

> 1. **Underscore Prefix**.  By convention, names that begin with an underscore

Reviewer 2: technicality: name mangling rules are confusing, let's stay away from them if we can (see https://docs.python.org/3/tutorial/classes.html#private-variables)
```suggestion
1. **Underscore Prefix**.  By convention, names that begin with a single underscore
```

These 3 means can be ambiguous and contradictory. Let's explore this point with
an example. Consider an example library with the following layout::

   src/
      __init__/
         __all__ = ["foo", "_bar"]

      foo/
         __all__ = ["_func"]
         def func():
            pass

         def _func():
            pass
      
      _bar/
         __all__ = ['bunc']
         def bunc():
            pass

         def _bunc():
            pass

      baz/
         __all__ = ['qux']

         def qux():
            pass

- Is ``foo`` public? Probably, but what if it's not documented?
- Is ``func`` public? It is not in ``foo.__all__``. 
- Is ``_func`` public?  It is prefixed with an underscore, but in ``foo.__all__``. 
- Is ``_bar`` public? It is prefixed with an underscore, but in ``src.__all__``.
- Is ``bunc`` public? It is in ``_bar.__all__``, but what if ``_bar`` is private?
- Is ``_bunc`` public? Probably not, but what if it's documented?
- Is ``baz`` public? It does not start with an underscore, but is not in ``src.__all__``.
- Is ``qux`` public? It does not start with an underscore and is in
  ``baz.__all__``, but what if ``baz`` is private?

> Astropy partially implements all three of the conventions: some private classes

Reviewer 3: >  astropy actually has a clear, basic definition of what is public, which is everything directly in a submodule like astropy.coordinates.

Do you know where this is written down?

> Astropy partially implements all three of the conventions: some private classes

Reviewer: I think it pretty obvious from the documentation - see the reference sections of `time`, `coordinates`, `timeseries` - though of course the moment I look at others, I see that `nddata` is not quite as good, and `modeling` has definite public modules (though clearly indicated as such).

I do stick with my main point that we actually have had very few cases where users were confused about where to import things from.

> Astropy partially implements all three of the conventions: some private classes

Reviewer 3: It was not too difficult to find counter-examples from [`astroquery`](https://github.com/search?q=repo%3Aastropy%2Fastroquery+%2Ffrom+astropy.table.table+import+Table%2F&type=code), [`specutils`](https://github.com/search?q=repo%3Aastropy%2Fspecutils+%2Ffrom+astropy.nddata.nduncertainty+import+StdDevUncertainty%2F&type=code), [`photutils`](https://github.com/search?q=repo%3Aastropy%2Fphotutils+%2Ffrom+astropy.nddata%5C.%5Ba-z%5D*%2F&type=code) and [`ccdproc`](https://github.com/search?q=repo%3Aastropy%2Fccdproc+%2Ffrom+astropy.units.quantity+import+Quantity%2F&type=code), and these are all coordinated packages.

> Astropy partially implements all three of the conventions: some private classes

Reviewer: Are you suggesting these should all be converted to a particular scheme too?Then we better get more feedback on this APE  :smiley_cat: I doubt I'm the only one who feels time is better spent elsewhere (and sadly this is not something that can be done without impacting everybody).

> Astropy partially implements all three of the conventions: some private classes

Reviewer 3: I presented a few counter-examples from coordinated packages to the claim 

> ...that we actually have had very few cases where users were confused about where to import things from.

We have more control over coordinated packages, so if those are being confused about importing then I'm guessing the confusion is even worse in non-coordinated packages.

communicating public vs internal symbols, which leads to confusion for users and
developers alike. This longstanding issue has steadily become more pressing with
the development of assistive code tools, like auto-fill in IDEs or more recently
with generative coding tools, like GitHub's CoPilot, which suggest code to users
based on a corpus of code examples.  When users unintentionally use internal
code, code tools will then suggest that internal code to other users. If we do
not have a uniform and systematic approach to communicating what is public vs
internal, then we cannot expect users and especially their tools to know what is
public vs internal.

Solution Criteria
-----------------

Having established that we need a uniform and systematic approach to public vs
internal interfaces, we now decide what that approach should be. The approach
should fulfill a few criteria:

1. Align closely with established conventions, so that it is familiar
   to users and developers, and hopefully detected by assistive tools.
2. Be unambiguous and easy to understand.
3. Be straightforward to implement and maintain, so that it is not a
   burden on developers.

> We start by considering established practice, for criteria 1, and how to ensure

Reviewer 2: ```suggestion
We start by considering established practice, for criterion 1, and how to ensure
```

introspection by both users and automated systems.

Finally, let's consider what `Python typing guide
<https://github.com/python/typing/blob/master/docs/source/libraries.rst#library-interface-public-and-private-symbols>`
adds.:: 

   - Symbols whose names begin with an underscore (but are not dunder names) are
     considered private.
   
   - A module can expose an ``__all__`` symbol at the module level that provides
     a list of names that are considered part of the interface. This overrides
     all other rules above, allowing imported symbols or symbols whose names
     begin with an underscore to be included in the interface.

This is consistent with PEP 8 and SciPy, but it does not mention the
documentation. The Python typing guide reinforces the idea that the ``__all__``
attribute is the authoritative source of what is public vs internal.
Importantly, the ``__all__`` attribute is a module-level attribute, so it only
applies to the module in which it is defined. We will refer to this as "locally"
public. This means that a module can define an ``__all__`` attribute but if the
module itself is not public, then anything in the ``__all__`` attribute cannot

> be publicly accessed from outside the module.

Reviewer 2: I'm not sure I understand this sentence. Does this imply that any private module should set `__all__ = []`, or does it mean that `__all__`, within a private module, defines an interface that's meant for internal use only ? (now that I wrote it down, the latter seems way more likely, but I think there's room to improve the phrasing)

Author: Yes, the latter. If you can't publicly navigate to a module, it doesn't matter what's "public" within that module. The full path must be public.
The issue with relying on this alone is what you discovered in that Issue (trying to find the one) where we have modules that look public (because they are not underscored) so it's possible to construct an import path that looks totally public, but actually isn't.

Reviewer 2: >  If you can't publicly navigate to a module, it doesn't matter what's "public" within that module.

I think that's what confused me; since it doesn't matter, why should we care what in a private module's `__all__` ? Wouldn't it be simpler to recommend/rule that `__all__` must be set to `[]` in private modules ?

> The issue with relying on this alone is what you discovered in that Issue (trying to find the one)

For reference, I believe you meant to link https://github.com/astropy/astropy/issues/15666

Reviewer 4: It might still be useful to be able to indicate in a private sub-module which things are meant to be importable in other parts of astropy, so ``__all__`` could be useful in that way rather than always setting to ``[]``.

Author: Agreed! This is demonstrated in lines 281-287

Author: In examples of good module layouts.

Author: @astrofrog, see https://github.com/astropy/astropy/pull/15109/files as an example doing what you're talking about. I agree, this is super useful.

4. That Astropy strongly recommends these rules for coordinated packages, and
   encourage affiliated packages to follow these rules as well.

**Rules for Public Interfaces:**

1. A symbol is public if all containing namespaces are public.
   A symbol is internal if any containing namespace is internal.

2. All modules must have an ``__all__`` attribute, even if it is empty. The
   ``__all__`` attribute specifies the public and internal interface of the
   module in which it is defined. Anything in ``__all__`` is *locally* public,
   including underscore-prefixed symbols. Anything not in ``__all__`` is
   internal in that module. The exception to this rule are modules that cannot
   have an ``__all__`` attribute, for example implicit namespace packages, which
   lack an ``__init__.py``. In these cases, the public interface is defined by
   the sub-rules:

   - A symbol is *locally* public if it does not start with an underscore, and
     internal otherwise, unless it is a dunder symbol (``__<...>__``).
   - Dunder symbols (``__<...>__``) have their own public vs internal rules,
     which are not determined in this APE.

3. Public symbols must not be prefixed with an underscore.

>    - Internal symbols need not be prefixed with an underscore, but it is often

Author: Perhaps this should be strengthened.
"Internal symbols must be prefixed with an underscore, unless the modules in which the symbol is used are all private, making the prefix redundant"

Reviewer 2: I'm in favour of this rephrasing !

Author: How about...

It is recommended, though not enforced, that internal symbols should be prefixed by a single underscore. An example of when the underscore is not necessary is for internal symbols defined in an internal module, since the prefix would be redundant (e.g. ``_bar`` in ``_foo.py`` can be just ``bar``).

Reviewer 2: sounds good to me, thank you !

     recommended. An example of when it is not needed are symbols defined in
     internal modules, making the prefix redundant.

4. Public symbols must be documented.

   - If an internal symbol is documented (which is not recommended), it must be

>      obviously, explicitly, (and preferably repeatedly) noted as internal.

Reviewer 2: "obvious" may sometimes give off a condescending vibe to the reader, let's avoid that
```suggestion
     unambiguously, explicitly, (and preferably repeatedly) noted as internal.
```

Let's consider an example::

>    src/

Author: Note to self: difficulties in public/private API definitions in https://github.com/astropy/astropy/pull/16519

      __init__.py::
         __all__ = ["foo", "spam"]

      foo.py::
         __all__ = ["func"]
         def func():
            pass

         def _func():
            pass
      
      _bar.py::
         __all__ = ['bunc']
         def bunc():
            pass

         def _bunc():
            pass

      baz.py::
         __all__ = ['qux']

         def qux():
            pass

      spam/  # implicit namespace package

         ham.py::
            __all__ = ['eggs']

            def eggs():
               pass

In this case, ``bar`` is public in ``foo``, and  ``foo`` is public in ``src``,

> so ``src.foo.func`` is public. This is public. In contrast, while ``bunc`` is

Reviewer 2: ```suggestion
so ``src.foo.func`` is public. In contrast, while ``bunc`` is
```

                    ``._bunc``              internal
         ``spam``                           public
                    ``.ham``                public
                                ``.eggs``   public
=======  =========  ==========  ==========  =========================

Branches and pull requests
--------------------------

N/A

Implementation
--------------

The process of adopting this APE will change Astropy's API. This APE formally
establishes Astropy's API, so what is meant by "change"?  There is already a *de
facto* API, which is what users *think* is the API.  This is determined
primarily by the documentation, but also by tab-completion within an interactive
session.  In the implementation of this APE to establish a *formal* API, we aim
to minimize changes to the *de facto* API. This will be accomplished in the
following 5 phases:

1. **Snapshot the documentation**. As of the adoption of the APE a snapshot of
   the documentation will be saved. This will be used to determine what is
   currently public and what is not. Until the APE's adoption is complete, this
   snapshot is authoritative, e.g. dictating what must be added and removed from

>    ``__all__``, for deciding what must undergo a deprecation process, etc.

Reviewer 2: I feel like we lack a prescription for what to do with symbols that are added post-snapshot here

Author: They should follow the rules! If it's public, make it so.
This is just saying we're not going to debate what already exists until we rectify the current ad-hoc standard with this one. But new things should follow the new standard.

Reviewer 2: Agreed, but I think I meant to suggest that it should be explicitly mentioned in the APE too.

2. **Add / update** ``__all__``. The ``__all__`` in each module will be updated
   to reflect phase 1. Any modules' missing ``__all__`` will have one added.

3. **Update the documentation** and **implement deprecations**.

   - Make sure every public symbol is in the documentation, unless it is being
     deprecated.
   - Deprecate any public symbol that is not intended to be public, but was made
     public as part of steps 1 and 2.
   - Add a SciPy-like section to the developer documentation explaining the
     public vs internal rules.
   - Fix links to always point to the public interface, not the internal
     interface. (This will presumably require a similar implementation as used
     in NumPy to change the ``__module__`` attribute of many objects.)
   - Add a reminder to the maintainer checklist that to be public a symbol must be
     in ``__all__`` and documented.
   - Clearly state if a documented object is actually private.

> 4. **Add a pre-commit bot**. This will ensure that the ``__all__`` attribute is

Reviewer 2: ?
```suggestion
4. **Add a pre-commit hook**. This will ensure that the ``__all__`` attribute is
```

Author: 👍

   directory for symbols in ``__all__``.

5. **Add prefixes**: Add prefixes to the top-most private symbols. For modules
   this makes all their contents private.

In Astropy core each phase will be implemented on a per-sub-package basis with
individual pull requests.

Backward compatibility
----------------------

This APE breaks backward compatibility in two ways:

1. Changes ``__all__`` in many modules. Many modules define an ``__all__`` but
   include symbols that are intended to be private because they should be
   imported from other modules (generally the top of the module). This will not
   break code that directly imports the symbols (as Python does not use
   ``__all__`` for this purpose), but it will break code that expects private
   symbols and uses ``import *``.
2. Adds prefixes to many symbols that are not public. This can be done in a
   backward compatible way by adding a ``__getattr__`` method to the module that
   raises a warning for any object that is not public. This will allow existing
   code to continue to work, while encouraging people to fix their code to use
   the public interface.

Alternatives
------------

**We do nothing:**

> This is the status quo.  It is not a good option because it is not solve the

Reviewer 2: ```suggestion
This is the status quo.  It is not a good option because it does not solve the
```

> issue.  The aforementioned problems of not i) knowing what is stable and

Reviewer 2: ```suggestion
issue.  The aforementioned problems of not knowing what is stable and
```

supported, and what is not, remain.

> **We allow** ``__all__`` **to be optional:**

Author: Note to self: IDK when they did this, but `scipy` now has underscore-prefixed module names everywhere. They fixed their public/private problem. e.g. https://github.com/scipy/scipy/tree/main/scipy/fft

This is not great.

The only time this might be good is when a module has dynamic symbols from a
`PEP 562 <https://peps.python.org/pep-0562/>`_ module-level ``__getattr__``

> method. However, if it truly dynamic then it cannot be statically analyzed, which

Reviewer 2: ```suggestion
method. However, if it is truly dynamic then it cannot be statically analyzed, which
```

is undesirable for other reasons. If it is not truly dynamic, then the "dynamic"
symbols can be added to ``__all__`` and an ``__init__.pyi`` file used to
communicate the public interface to static analyzers.

**We make the documentation the authoritative definition of the API.**

> Good code is well documented in the code. Beyond code comments and docstrings, what is

Reviewer 2: ?
```suggestion
Good code is well documented. Beyond code comments and docstrings, what is
```

Author: No, I mean in the code itself. E.g. we build documentation from code docstrings because we want the docstrings to be available in the code itself. If that weren't important all our docstrings would just be in the docs.

Reviewer 2: Oh, now I see what you meant !

public versus internal is an important aspect of the code, and should be
communicated in the code itself. It is vitally important that the code and
user-facing documentation are consistent, and the user-facing documentation
should reflect the code.

**We make tab-completion the authoritative definition of the API.**

This is not a good option because it does not work with implicit namespace
packages nor `PEP 562 <https://peps.python.org/pep-0562/>`_ module-level

> ``__getattr__`` methods.

Reviewer 2: just curious, is this mechanism actually used in astropy or in any affiliated package you know of ?

Author: - Implicit namespace packages: not any affiliated package I know, though I have some personally. Super useful. I wish tab-complete were better.
- ``__getattr__``, yeah in a couple places! E.g in `astropy.cosmology.realizations`. Also check out https://github.com/scientific-python/lazy_loader. I would very much like for Astropy to implement lazy loading and speed up initialization time.

Reviewer 2: Thanks ! My question was specifically about implicit namespace packages as I have first hand experience with module-level `__getattr__` and I know it's extremely useful. Nonetheless, I'm glad my question leaked to this topic too because I'm also interested in lazy loading and startup time. Actually astropy already has it a lot better than a lot of packages I've seen, but I'm sure we can improve it still. Anyway this is a digression, I just wanted to let you know that I'm interested in that topic as well. Feel free to resolve this thread !
